### ISSUE TITLE
Not clear which layers are being trained

### ISSUE THREAD

**Issue by peacej** [UKPLab/sentence-transformers#129]
In your paper, you state "The purpose of SBERT sentence embeddings are not to be used for transfer learning for other tasks. Here, we think fine-tuning BERT as described by Devlin et al. (2018) for new tasks is the more suitable method, as it updates all layers of the BERT network."

The above is implying that the `SentenceTransformer.fit()` method does not update all layers of BERT (or XLNet, etc.) . However, there is no code in this repo that partially freezes any layers when training or anything like that.  Could you kindly clarify how you're controlling which layers are being trained (other than the parameters in the loss modules, which are clearly being trained) ? Sorry if this is a stupid question.

**Comment by nreimers** [comment]
All layers are fine tuned in the fit method.

Sentence Embeddings are often advertised to be used for transfer learning. You map you sentence to a vector and the train e. g. a logistic regression classifier on top for example for sentiment classification. There, you would only update the classifier but not the sentence embeddings method.

For these supervised cases it makes more sense to ditch sentence embeddings and to fine tune BERT directly.

Best regards 
Nils Reimers

**Comment by peacej** [comment]
Thanks for your kind explanation~ 

For supervised (transfer learning) cases, why do you think "it makes more sense to ditch sentence embeddings and to fine tune BERT directly" ?  Your paper gives the reason that fine-tuning BERT directly is more suitable because "it updates all layers of the BERT network," however you just confirmed that `SentenceTransformer.fit()` also updates all layers of BERT.  Furthermore, you even provide an example where you do supervised transfer learning (https://github.com/UKPLab/sentence-transformers/blob/master/examples/training_stsbenchmark_continue_training.py). 

For my use case (transfer learning), I would like to train parameters for not just the final classifier layer but also some or all of the BERT layers because I believe my domain is specialized (and has the tendency to use certain words/phrases in a certain way). I do have lots of training labels which can be used with your loss modules such as `SoftmaxLoss()`. But I guess you're saying that I shouldn't use this `sentence-transformers` package?

**Comment by kevinmandich** [comment]
@peacej I have a similar situation with a specialized domain. What I do is a two-piece training process:

1 - Finetune BERT in an unsupervised fashion on my domain text (via Huggingface's transformers repo)
2 - Fit a SentenceTransformer on labeled pairs, using the weights from 1 as the base model.

**Comment by peacej** [comment]
@kevinmandich that sounds like a fantastic idea, I'll try that!

**Comment by nreimers** [comment]
@peacej 
The statement was in reference to sentence classification tasks, like sentiment classification. These tasks are evaluated for example with SentEval.

There (in SentEval), the assumption is:
Sentencen ---Fixed sentence embedding method ---> Vector --- Log reg. --> Label


In that SentEval scenario, only the logistic regression classifier is updated. The sentence embedding method is fixed.

If you have a sentence classification task, like sentiment classification, I think this setup doesn't make sense. There, it is better to fine-tune bert:
Sentence ---- BERT ---> Label

Sentence embedding methods make sense for unsupervised tasks and tasks where you use cosine similarity / manhatten- / euclidian-distance for retrieval etc.


I hope this makes it a bit more clear.

Best
Nils Reimers

**Comment by peacej** [comment]
Thanks Nils! I actually have a hybrid situation where I need to use cosine similarity for retrieval from a large number of pre-computed candidate vectors given a query vector, but I'm hoping that the retrieved vector matched to the query vector will lead to a positive result (e.g. a click by a user), so that's kind of like binary classification.

FYI, I tried using different loss classes that you provided.  When training with `SoftmaxLoss`, it performs well on classification (when including the final linear layer, denoted as W_t in your paper) but poorly when just using cosine similarity (which excludes the final linear layer). So it seems the model depends heavily on the final linear layer for good classification. However, when training using TripletLoss, just using cosine/euclidean distance performs decently on classification, which I guess makes sense since training also uses cosines/euclidean distance and has no dependency on an additional linear layer. I'll also try fine-tuning BERT directly as you suggested.

Anyway, I'll close this issue for now to keep your repo clean, but I'd be interested in any more related discussions.
